### ISSUE TITLE
[Question] Which FA interface support custom `attention_bias`?

### ISSUE THREAD

**Issue by tutel-agent** [Dao-AILab/flash-attention#1805]
I'd like to call an efficient `flash_attn_interface(.., attn_bias=custom_input, ..)` where attn_bias is neither None nor casual_mask. Is there any option?

**Comment by tridao** [comment]
Sliding window, that's about it.

**Comment by tutel-agent** [comment]
Is it `window_size`? That also seems to be a kind of specialized masking layout, while I still need to configure my own masking layout in detail, for example:

<img width="548" height="526" alt="Image" src="https://github.com/user-attachments/assets/6eb93acf-ae18-4c1f-b794-84d4b0dd5d72" />

**Comment by tridao** [comment]
yes windowsize is used to specify sliding window. The mask you mentioned isn't supported, you can use FlexAttention for that.

**Comment by tutel-agent** [comment]
Thanks! But FlexAttention isn't efficient enough, and it is also inflexible to do fine-grain customization.

I'd like to do `softmax(Q(.., seq=4, ..) @ K(.., seq=8, ..)) @ V(.., seq=8, ...)`, so the attention_mask (of shape [4, 8]) is expected to be nearly causal:

```txt
T T T T T F F F
T T T T T T F F
T T T T T T T F
T T T T T T T T
```

Does `is_causal` or `window_size` support that masking layout?
If not, which code line (based on FA2) can I do modification to support it?

**Comment by tridao** [comment]
F means keep and T means masked out?
Then if you remove the last token of K and V it would be exactly causal?

**Comment by tutel-agent** [comment]
Sorry for a mistake which has been fixed inline: the query sequence length should be shorter than KV sequence length, and T(rue) stands for `+ attn_bias(0.0)` (indicating keep) while F(alse) stands for `+ attn_bias(-inf)` (indicating masked out).

Like this:

```txt
T T T T T F F F
T T T T T T F F
T T T T T T T F
T T T T T T T T
```

I think when query sequence length == kv sequence length, above layout will be exactly identical to standard `is_causal`.

**Comment by tridao** [comment]
This is also exactly causal 
https://github.com/Dao-AILab/flash-attention/blob/3c51f15dc04c05e97cae1cfbd494e1f02962516a/flash_attn/flash_attn_interface.py#L1394
